src/Stratis.Bitcoin/Utilities/AsyncQueue.cs

```diff
@@ -93,11 +113,95 @@ public void Enqueue(T item)
             }
         }
 
+        /// <summary>
+        /// Dequeues an item from the queue if there is one. 
+        /// If the queue is empty, the method waits until an item is available.
+        /// </summary>
+        /// <param name="cancellation">Cancellation token that allows aborting the wait if the queue is empty.</param>
+        /// <returns>Dequeued item from the queue.</returns>
+        /// <exception cref="OperationCanceledException">Thrown when the cancellation token is triggered or when the queue is disposed.</exception>
+        /// <exception cref="InvalidOperationException">Thrown if this method is called on a queue that operates in callback mode.</exception>
+        public async Task<T> DequeueAsync(CancellationToken cancellation = default(CancellationToken))
+        {
+            if (this.callbackMode)
+                throw new InvalidOperationException($"{nameof(DequeueAsync)} called on queue in callback mode.");
+
+            // Increment the counter so that the queue's cancellation source is not disposed when we are using it.
+            Interlocked.Increment(ref this.unfinishedDequeueCount);
+
+            try
+            {
+                if (this.disposed)
+                    throw new OperationCanceledException();
+
+                // First check if an item is available. If it is, just return it.
+                T item;
+                if (this.TryDequeue(out item))
+                    return item;
+
+                // If the queue is empty, we need to wait until there is an item available.
+                using (var cancellationSource = CancellationTokenSource.CreateLinkedTokenSource(cancellation, this.cancellationTokenSource.Token))
```

Comment:
Why use a linked source at no point you are cancelling the cancellation.

Reply:
i use it in tests, do you want me to remove it and explicitly use none token there in test?

Reply:
Hmm to add a linked cancellation for tests sounds wrong, I think there is a performance hit for creating a linked token and potentially you might leak handles if not disposed correctly.

If those are not concerns (as we do dispose the linked token) then perhaps its not a big deal but we still don't have a usecase, I would prefer to remove it but wont insist.

Reply:
it is not linked for the tests, the link will be there always, for the tests there is the default value, so what i'm saying is that i can remove the default value, but the link will still be there

there is no leak, we are using `using`, so no leaks

Reply:
What about perf hit? we are creating a linked cancellation token for every loop in the iteration.

Reply:
the cancellation is needed
perf hit is minimal, this dequeue operation is supposed to be blocking, so most of the time you wait for the item to arrive

Reply:
Good point ok I will take that.